graphql/websocket.go

```diff
@@ -139,22 +143,31 @@ func (w *webSocketClient) forwardWebSocketData(message []byte) error {
 	if wsMsg.ID == "" { // e.g. keep-alive messages
 		return nil
 	}
-	w.subscriptions.Lock()
-	defer w.subscriptions.Unlock()
-	sub, success := w.subscriptions.map_[wsMsg.ID]
-	if !success {
+
+	w.Lock()
+	sub := w.subscriptions.get(wsMsg.ID)
+	if sub == nil {
+		w.Unlock()
 		return fmt.Errorf("received message for unknown subscription ID '%s'", wsMsg.ID)
 	}
-	if sub.hasBeenUnsubscribed {
+	if sub.closed {
+		// Already closed, ignore message
+		w.Unlock()
 		return nil
 	}
+
 	if wsMsg.Type == webSocketTypeComplete {
-		sub.hasBeenUnsubscribed = true
-		w.subscriptions.map_[wsMsg.ID] = sub
-		reflect.ValueOf(sub.interfaceChan).Close()
+		// Server is telling us the subscription is complete
+		w.subscriptions.closeChannel(wsMsg.ID)
+		w.subscriptions.delete(wsMsg.ID)
+		w.Unlock()
 		return nil
 	}
 
+	// Forward the data to the subscription channel.
+	// We release the lock while calling the forward function to avoid holding
+	// the lock while doing potentially slow user code.
+	w.Unlock()
```

Comment:
In this case, I do agree that having to remember to unlock in all the exit paths is fragile.  It may make sense to pull all this code out into its own function so you can use `defer` to close.